// @figspec/components/esm/es2019/ui/ui.js
import { className, el } from "../dom.js";
import { compute, Signal } from "../signal.js";
import { canvas, info, preferences, isIdle, isSetupError, isInfo, isPreferences, } from "../state.js";
import { empty } from "./empty/empty.js";
import { fullscreenPanel } from "./fullscreenPanel/fullscreenPanel.js";
import { inspectorPanel } from "./inspectorPanel/inspectorPanel.js";
import { menuBar } from "./menuBar/menuBar.js";
import { preferencesPanel } from "./preferencesPanel/preferencesPanel.js";
import { snackbar } from "./snackbar/snackbar.js";
export const styles = /* css */ `
  .ui-root {
    border-radius: inherit;
  }
`;
const SNACKBAR_LIFETIME = 3000;
export function ui({ state: $state, preferences: $preferences, infoContents: createinfoContents, frameCanvas: createFrameCanvas, menuSlot: createMenuSlot, caller, }) {
    const $snackbar = new Signal(null);
    return compute(() => {
        const s = $state.get();
        if (isIdle(s)) {
            if (caller === "file") {
                return empty({
                    title: ["No Figma file"],
                    body: [
                        el("p", [], [
                            "Both Figma file data and rendered images are missing. ",
                            "Please provide those in order to start Figspec File Viewer. ",
                        ]),
                    ],
                });
            }
            return empty({
                title: ["No Figma frame"],
                body: [
                    el("p", [], [
                        "Both frame data and rendered image are missing. ",
                        "Please provide those in order to start Figspec Frame Viewer. ",
                    ]),
                ],
            });
        }
        if (isSetupError(s)) {
            return empty({
                title: ["Failed to render Figma ", caller],
                body: [
                    el("p", [], ["Couldn't render the Figma ", caller, " due to an error."]),
                    el("pre", [], [s.error.message, "\n\n", s.error.stack]),
                ],
            });
        }
        const $loadedState = new Signal(canvas);
        const $selected = new Signal(null);
        const frameCanvas = createFrameCanvas(s.data, $selected, $loadedState, $snackbar);
        const perState = compute(() => {
            const loadedState = $loadedState.get();
            if (isInfo(loadedState)) {
                return fullscreenPanel({
                    body: [createinfoContents(s.data)],
                    onClose() {
                        $loadedState.set(canvas);
                    },
                });
            }
            if (isPreferences(loadedState)) {
                return fullscreenPanel({
                    body: [preferencesPanel({ preferences: $preferences })],
                    onClose() {
                        $loadedState.set(canvas);
                    },
                });
            }
            return el("div", [], [
                menuBar({
                    slot: [createMenuSlot === null || createMenuSlot === void 0 ? void 0 : createMenuSlot(s.data)],
                    onOpenInfo() {
                        $loadedState.set(info);
                    },
                    onOpenPreferences() {
                        $loadedState.set(preferences);
                    },
                }),
                inspectorPanel({
                    selected: $selected,
                    preferences: $preferences,
                    snackbar: $snackbar,
                    onOpenPreferencesPanel() {
                        $loadedState.set(preferences);
                    },
                }),
            ]);
        });
        const layer = el("div", [className("ui-root")], [
            frameCanvas,
            perState,
            snackbar({ signal: $snackbar, lifetimeMs: SNACKBAR_LIFETIME }),
        ]);
        return layer;
    });
}
