// @figspec/components/esm/es2019/FigspecFileViewer.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FigspecFileViewer_link, _FigspecFileViewer_resp, _FigspecFileViewer_images, _FigspecFileViewer_canvases, _FigspecFileViewer_selectedCanvasId, _FigspecFileViewer_state, _FigspecFileViewer_givenPreferences, _FigspecFileViewer_preferences, _FigspecFileViewer_ui;
import { attr, el } from "./dom.js";
import * as figma from "./figma.js";
import { FrameCanvas } from "./FrameCanvas/FrameCanvas.js";
import { defaultPreferenecs, isEqual as isEqualPreferences, } from "./preferences.js";
import { compute, effect, Signal } from "./signal.js";
import { styles } from "./styles.js";
import * as state from "./state.js";
import { ui } from "./ui/ui.js";
import { infoItems } from "./ui/infoItems/infoItems.js";
import { selectBox } from "./ui/selectBox/selectBox.js";
export class FigspecFileViewer extends HTMLElement {
    set preferences(value) {
        __classPrivateFieldSet(this, _FigspecFileViewer_givenPreferences, value, "f");
        __classPrivateFieldGet(this, _FigspecFileViewer_preferences, "f").set({ ...value });
    }
    get preferences() {
        return __classPrivateFieldGet(this, _FigspecFileViewer_preferences, "f").once();
    }
    constructor() {
        super();
        _FigspecFileViewer_link.set(this, new Signal(null));
        _FigspecFileViewer_resp.set(this, new Signal(null));
        _FigspecFileViewer_images.set(this, new Signal(null));
        _FigspecFileViewer_canvases.set(this, compute(() => {
            const map = new Map();
            const resp = __classPrivateFieldGet(this, _FigspecFileViewer_resp, "f").get();
            if (!resp) {
                return map;
            }
            for (const canvas of figma.getCanvases(resp.document)) {
                map.set(canvas.id, canvas);
            }
            return map;
        }));
        _FigspecFileViewer_selectedCanvasId.set(this, new Signal(null));
        _FigspecFileViewer_state.set(this, compute(() => {
            const resp = __classPrivateFieldGet(this, _FigspecFileViewer_resp, "f").get();
            const images = __classPrivateFieldGet(this, _FigspecFileViewer_images, "f").get();
            if (!resp && !images) {
                return state.idle;
            }
            if (!images) {
                return state.setupError(new Error("Rendered image set is required"));
            }
            if (!resp) {
                return state.setupError(new Error("Returned result of Get File API is required"));
            }
            const canvases = __classPrivateFieldGet(this, _FigspecFileViewer_canvases, "f").get();
            if (!canvases.size) {
                return state.setupError(new Error("No node has type=CANVAS."));
            }
            return state.loaded([resp, images, canvases]);
        }));
        _FigspecFileViewer_givenPreferences.set(this, { ...defaultPreferenecs });
        _FigspecFileViewer_preferences.set(this, new Signal({ ...__classPrivateFieldGet(this, _FigspecFileViewer_givenPreferences, "f") }));
        _FigspecFileViewer_ui.set(this, ui({
            caller: "file",
            state: __classPrivateFieldGet(this, _FigspecFileViewer_state, "f"),
            preferences: __classPrivateFieldGet(this, _FigspecFileViewer_preferences, "f"),
            infoContents: ([resp, , canvases]) => {
                return infoItems([
                    {
                        label: "Filename",
                        content: [resp.name],
                    },
                    {
                        label: "Last modified",
                        content: [new Date(resp.lastModified).toLocaleString()],
                    },
                    compute(() => {
                        const link = __classPrivateFieldGet(this, _FigspecFileViewer_link, "f").get();
                        if (!link) {
                            return null;
                        }
                        return {
                            label: "File link",
                            content: [
                                el("a", [
                                    attr("href", link),
                                    attr("target", "_blank"),
                                    attr("rel", "noopener"),
                                ], [link]),
                            ],
                        };
                    }),
                    {
                        label: "Number of canvases",
                        content: [canvases.size.toString(10)],
                    },
                ]);
            },
            menuSlot: ([, , canvases]) => {
                return compute(() => {
                    return selectBox({
                        value: compute(() => { var _a; return (_a = __classPrivateFieldGet(this, _FigspecFileViewer_selectedCanvasId, "f").get()) !== null && _a !== void 0 ? _a : ""; }),
                        options: Array.from(canvases).map(([, canvas]) => el("option", [attr("value", canvas.id)], [canvas.name])),
                        onChange: (value) => {
                            if (canvases.has(value)) {
                                __classPrivateFieldGet(this, _FigspecFileViewer_selectedCanvasId, "f").set(value);
                            }
                        },
                    });
                });
            },
            frameCanvas: ([, images, canvases], $selected, $loadedState, $snackbar) => {
                const frameCanvas = new FrameCanvas(__classPrivateFieldGet(this, _FigspecFileViewer_preferences, "f"), $selected, $snackbar);
                effect(() => {
                    $selected.set(null);
                    const currentId = __classPrivateFieldGet(this, _FigspecFileViewer_selectedCanvasId, "f").get();
                    if (typeof currentId !== "string") {
                        return;
                    }
                    const node = canvases.get(currentId);
                    if (!node) {
                        return;
                    }
                    frameCanvas.render([node], images);
                    return () => {
                        frameCanvas.clear();
                    };
                });
                let isFirstRun = true;
                effect(() => {
                    const selected = $selected.get();
                    if (isFirstRun) {
                        isFirstRun = false;
                        return;
                    }
                    this.dispatchEvent(new CustomEvent("nodeselect", {
                        detail: {
                            node: selected,
                        },
                    }));
                });
                effect(() => {
                    if (!state.isCanvas($loadedState.get())) {
                        return;
                    }
                    frameCanvas.connectedCallback();
                    return () => {
                        frameCanvas.disconnectedCallback();
                    };
                });
                return frameCanvas.container;
            },
        }));
        const shadow = this.attachShadow({ mode: "open" });
        const style = document.createElement("style");
        style.textContent += styles;
        style.textContent += FrameCanvas.styles;
        shadow.appendChild(style);
        effect(() => {
            for (const [id] of __classPrivateFieldGet(this, _FigspecFileViewer_canvases, "f").get()) {
                __classPrivateFieldGet(this, _FigspecFileViewer_selectedCanvasId, "f").set(id);
                return () => {
                    __classPrivateFieldGet(this, _FigspecFileViewer_selectedCanvasId, "f").set(null);
                };
            }
        });
        effect(() => {
            const node = __classPrivateFieldGet(this, _FigspecFileViewer_ui, "f").get();
            shadow.appendChild(node);
            return () => {
                shadow.removeChild(node);
            };
        });
        // Emit `preferencesupdate` event on preferences updates
        effect(() => {
            const preferences = __classPrivateFieldGet(this, _FigspecFileViewer_preferences, "f").get();
            if (!isEqualPreferences(__classPrivateFieldGet(this, _FigspecFileViewer_givenPreferences, "f"), preferences)) {
                this.dispatchEvent(new CustomEvent("preferencesupdate", {
                    detail: { preferences },
                }));
            }
        });
    }
    set link(link) {
        __classPrivateFieldGet(this, _FigspecFileViewer_link, "f").set(link);
    }
    get link() {
        return __classPrivateFieldGet(this, _FigspecFileViewer_link, "f").once();
    }
    get renderedImages() {
        const map = __classPrivateFieldGet(this, _FigspecFileViewer_images, "f").once();
        if (!map) {
            return null;
        }
        return Object.fromEntries(map.entries());
    }
    set renderedImages(set) {
        const map = new Map();
        for (const nodeId in set) {
            map.set(nodeId, set[nodeId]);
        }
        __classPrivateFieldGet(this, _FigspecFileViewer_images, "f").set(map);
    }
    get apiResponse() {
        return __classPrivateFieldGet(this, _FigspecFileViewer_resp, "f").once();
    }
    set apiResponse(resp) {
        if (!resp) {
            return;
        }
        __classPrivateFieldGet(this, _FigspecFileViewer_resp, "f").set(resp);
    }
    attributeChangedCallback(name, oldValue, newValue) {
        if (newValue === oldValue) {
            return;
        }
        switch (name) {
            case "link": {
                __classPrivateFieldGet(this, _FigspecFileViewer_link, "f").set(newValue || null);
                return;
            }
        }
    }
}
_FigspecFileViewer_link = new WeakMap(), _FigspecFileViewer_resp = new WeakMap(), _FigspecFileViewer_images = new WeakMap(), _FigspecFileViewer_canvases = new WeakMap(), _FigspecFileViewer_selectedCanvasId = new WeakMap(), _FigspecFileViewer_state = new WeakMap(), _FigspecFileViewer_givenPreferences = new WeakMap(), _FigspecFileViewer_preferences = new WeakMap(), _FigspecFileViewer_ui = new WeakMap();
FigspecFileViewer.observedAttributes = ["link"];
