// @figspec/components/esm/es2019/FigspecFrameViewer.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FigspecFrameViewer_link, _FigspecFrameViewer_resp, _FigspecFrameViewer_image, _FigspecFrameViewer_state, _FigspecFrameViewer_givenPreferences, _FigspecFrameViewer_preferences, _FigspecFrameViewer_ui;
import { attr, el } from "./dom.js";
import { FrameCanvas } from "./FrameCanvas/FrameCanvas.js";
import { defaultPreferenecs, isEqual as isEqualPreferences, } from "./preferences.js";
import { compute, effect, Signal } from "./signal.js";
import { styles } from "./styles.js";
import * as state from "./state.js";
import { ui } from "./ui/ui.js";
import { infoItems } from "./ui/infoItems/infoItems.js";
export class FigspecFrameViewer extends HTMLElement {
    set preferences(value) {
        __classPrivateFieldSet(this, _FigspecFrameViewer_givenPreferences, value, "f");
        __classPrivateFieldGet(this, _FigspecFrameViewer_preferences, "f").set({ ...value });
    }
    get preferences() {
        return __classPrivateFieldGet(this, _FigspecFrameViewer_preferences, "f").once();
    }
    constructor() {
        super();
        _FigspecFrameViewer_link.set(this, new Signal(null));
        _FigspecFrameViewer_resp.set(this, new Signal(null));
        _FigspecFrameViewer_image.set(this, new Signal(null));
        _FigspecFrameViewer_state.set(this, compute(() => {
            const resp = __classPrivateFieldGet(this, _FigspecFrameViewer_resp, "f").get();
            const image = __classPrivateFieldGet(this, _FigspecFrameViewer_image, "f").get();
            if (!resp && !image) {
                return state.idle;
            }
            if (!image) {
                return state.setupError(new Error("Image file URI is required"));
            }
            if (!resp) {
                return state.setupError(new Error("Returned result of Get File Nodes API is required"));
            }
            const node = findMainNode(resp);
            if (!node) {
                return state.setupError(new Error("No renderable node is found in the data"));
            }
            return state.loaded([resp, node, image]);
        }));
        _FigspecFrameViewer_givenPreferences.set(this, { ...defaultPreferenecs });
        _FigspecFrameViewer_preferences.set(this, new Signal({ ...__classPrivateFieldGet(this, _FigspecFrameViewer_givenPreferences, "f") }));
        _FigspecFrameViewer_ui.set(this, ui({
            caller: "frame",
            state: __classPrivateFieldGet(this, _FigspecFrameViewer_state, "f"),
            preferences: __classPrivateFieldGet(this, _FigspecFrameViewer_preferences, "f"),
            infoContents: ([resp]) => {
                return infoItems([
                    {
                        label: "Filename",
                        content: [resp.name],
                    },
                    {
                        label: "Last modified",
                        content: [new Date(resp.lastModified).toLocaleString()],
                    },
                    compute(() => {
                        const link = __classPrivateFieldGet(this, _FigspecFrameViewer_link, "f").get();
                        if (!link) {
                            return null;
                        }
                        return {
                            label: "Frame link",
                            content: [
                                el("a", [
                                    attr("href", link),
                                    attr("target", "_blank"),
                                    attr("rel", "noopener"),
                                ], [link]),
                            ],
                        };
                    }),
                ]);
            },
            frameCanvas: ([, node, image], $selected, $loadedState, $snackbar) => {
                const frameCanvas = new FrameCanvas(__classPrivateFieldGet(this, _FigspecFrameViewer_preferences, "f"), $selected, $snackbar);
                frameCanvas.render([node], new Map([[node.id, image]]));
                let isFirstRun = true;
                effect(() => {
                    const selected = $selected.get();
                    if (isFirstRun) {
                        isFirstRun = false;
                        return;
                    }
                    this.dispatchEvent(new CustomEvent("nodeselect", {
                        detail: {
                            node: selected,
                        },
                    }));
                });
                effect(() => {
                    if (!state.isCanvas($loadedState.get())) {
                        return;
                    }
                    frameCanvas.connectedCallback();
                    return () => {
                        frameCanvas.disconnectedCallback();
                    };
                });
                effect(() => {
                    return () => {
                        frameCanvas.clear();
                    };
                });
                return frameCanvas.container;
            },
        }));
        const shadow = this.attachShadow({ mode: "open" });
        const styleEl = document.createElement("style");
        styleEl.textContent += styles;
        styleEl.textContent += FrameCanvas.styles;
        shadow.appendChild(styleEl);
        effect(() => {
            const node = __classPrivateFieldGet(this, _FigspecFrameViewer_ui, "f").get();
            shadow.appendChild(node);
            return () => {
                shadow.removeChild(node);
            };
        });
        // Emit `preferencesupdate` event on preferences updates
        effect(() => {
            const preferences = __classPrivateFieldGet(this, _FigspecFrameViewer_preferences, "f").get();
            if (!isEqualPreferences(__classPrivateFieldGet(this, _FigspecFrameViewer_givenPreferences, "f"), preferences)) {
                this.dispatchEvent(new CustomEvent("preferencesupdate", {
                    detail: { preferences },
                }));
            }
        });
    }
    get link() {
        return __classPrivateFieldGet(this, _FigspecFrameViewer_link, "f").once();
    }
    set link(link) {
        __classPrivateFieldGet(this, _FigspecFrameViewer_link, "f").set(link);
    }
    get renderedImage() {
        return __classPrivateFieldGet(this, _FigspecFrameViewer_image, "f").once();
    }
    set renderedImage(uri) {
        if (uri) {
            __classPrivateFieldGet(this, _FigspecFrameViewer_image, "f").set(uri);
        }
    }
    get apiResponse() {
        return __classPrivateFieldGet(this, _FigspecFrameViewer_resp, "f").once();
    }
    set apiResponse(resp) {
        if (!resp) {
            return;
        }
        __classPrivateFieldGet(this, _FigspecFrameViewer_resp, "f").set(resp);
    }
    attributeChangedCallback(name, oldValue, newValue) {
        if (newValue === oldValue) {
            return;
        }
        switch (name) {
            case "link": {
                __classPrivateFieldGet(this, _FigspecFrameViewer_link, "f").set(newValue || null);
                return;
            }
        }
    }
}
_FigspecFrameViewer_link = new WeakMap(), _FigspecFrameViewer_resp = new WeakMap(), _FigspecFrameViewer_image = new WeakMap(), _FigspecFrameViewer_state = new WeakMap(), _FigspecFrameViewer_givenPreferences = new WeakMap(), _FigspecFrameViewer_preferences = new WeakMap(), _FigspecFrameViewer_ui = new WeakMap();
FigspecFrameViewer.observedAttributes = ["link"];
function findMainNode(resp) {
    for (const key in resp.nodes) {
        const node = resp.nodes[key];
        switch (node.document.type) {
            case "CANVAS":
            case "FRAME":
            case "GROUP":
            case "COMPONENT":
            case "COMPONENT_SET": {
                return node.document;
            }
        }
    }
    return null;
}
