// @figspec/components/esm/es2019/FrameCanvas/BoundingBoxMeasurement.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _BoundingBoxMeasurement_minX, _BoundingBoxMeasurement_maxX, _BoundingBoxMeasurement_minY, _BoundingBoxMeasurement_maxY;
/**
 * Measure bounding box for nodes.
 */
export class BoundingBoxMeasurement {
    constructor() {
        _BoundingBoxMeasurement_minX.set(this, Infinity);
        _BoundingBoxMeasurement_maxX.set(this, -Infinity);
        _BoundingBoxMeasurement_minY.set(this, Infinity);
        _BoundingBoxMeasurement_maxY.set(this, -Infinity);
    }
    /**
     * Add a node to the measurement.
     */
    addNode(node) {
        if (node.visible === false) {
            return;
        }
        const box = node.absoluteRenderBounds || node.absoluteBoundingBox;
        __classPrivateFieldSet(this, _BoundingBoxMeasurement_minX, Math.min(__classPrivateFieldGet(this, _BoundingBoxMeasurement_minX, "f"), box.x), "f");
        __classPrivateFieldSet(this, _BoundingBoxMeasurement_maxX, Math.max(__classPrivateFieldGet(this, _BoundingBoxMeasurement_maxX, "f"), box.x + box.width), "f");
        __classPrivateFieldSet(this, _BoundingBoxMeasurement_minY, Math.min(__classPrivateFieldGet(this, _BoundingBoxMeasurement_minY, "f"), box.y), "f");
        __classPrivateFieldSet(this, _BoundingBoxMeasurement_maxY, Math.max(__classPrivateFieldGet(this, _BoundingBoxMeasurement_maxY, "f"), box.y + box.height), "f");
    }
    /**
     * Returns a bounding box for added nodes.
     */
    measure() {
        return {
            x: Number.isFinite(__classPrivateFieldGet(this, _BoundingBoxMeasurement_minX, "f")) ? __classPrivateFieldGet(this, _BoundingBoxMeasurement_minX, "f") : NaN,
            y: Number.isFinite(__classPrivateFieldGet(this, _BoundingBoxMeasurement_minY, "f")) ? __classPrivateFieldGet(this, _BoundingBoxMeasurement_minY, "f") : NaN,
            width: Number.isFinite(__classPrivateFieldGet(this, _BoundingBoxMeasurement_maxX, "f")) && Number.isFinite(__classPrivateFieldGet(this, _BoundingBoxMeasurement_minX, "f"))
                ? __classPrivateFieldGet(this, _BoundingBoxMeasurement_maxX, "f") - __classPrivateFieldGet(this, _BoundingBoxMeasurement_minX, "f")
                : NaN,
            height: Number.isFinite(__classPrivateFieldGet(this, _BoundingBoxMeasurement_maxY, "f")) && Number.isFinite(__classPrivateFieldGet(this, _BoundingBoxMeasurement_minY, "f"))
                ? __classPrivateFieldGet(this, _BoundingBoxMeasurement_maxY, "f") - __classPrivateFieldGet(this, _BoundingBoxMeasurement_minY, "f")
                : NaN,
        };
    }
}
_BoundingBoxMeasurement_minX = new WeakMap(), _BoundingBoxMeasurement_maxX = new WeakMap(), _BoundingBoxMeasurement_minY = new WeakMap(), _BoundingBoxMeasurement_maxY = new WeakMap();
