// @figspec/components/esm/es2019/ui/infoItems/infoItems.js
import { attr, className, el } from "../../dom.js";
import { compute, Signal } from "../../signal.js";
export const styles = /* css */ `
  .ii-list {
    padding: 0;
    margin: 0;
    user-select: text;
  }

  .ii-label {
    padding: 0;
    margin: 0;
    margin-bottom: var(--spacing_-5);
    font-size: calc(var(--font-size) * 0.8);
    font-weight: bold;

    color: var(--subtle-fg);
  }

  .ii-content {
    padding: 0;
    margin: 0;
    margin-bottom: var(--spacing_3);
    font-size: var(--font-size);
    font-weight: normal;

    color: var(--fg);
  }
`;
export function infoItems(items) {
    return el("dl", [className("ii-list")], items
        .map((item) => {
        if (!item) {
            return [];
        }
        if (item instanceof Signal) {
            return [map(label, item), map(content, item)];
        }
        return [label(item), content(item)];
    })
        .flat());
}
function map(f, s) {
    return compute(() => {
        const v = s.get();
        if (v === null) {
            return null;
        }
        return f(v);
    });
}
function label(item) {
    return el("dt", [className("ii-label")], [item.label, el("span", [attr("aria-hidden", "true")], [":"])]);
}
function content(item) {
    return el("dd", [className("ii-content")], item.content);
}
